lib/srv/db/redis/engine.go

```diff
@@ -261,26 +248,109 @@ func (e *Engine) getNewClientFn(ctx context.Context, sessionCtx *common.Session)
+	// See: https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/auth-iam.html#auth-iam-limits
+	// So we must check that the database supports IAM auth and that the
+	// ElastiCache user has IAM auth enabled.
+	case e.isAWSIAMAuthSupported(ctx, sessionCtx):
+		credFetchFn := elasticacheIAMTokenFetchFunc(sessionCtx, e.Auth)
+		return fetchCredentialsOnConnect(e.Context, sessionCtx, e.Audit, credFetchFn)
 
 	default:
 		return nil
 	}
 }
 
+// isAWSIAMAuthSupported returns whether AWS IAM auth is supported for the
+// database and the database user.
+func (e *Engine) isAWSIAMAuthSupported(ctx context.Context, sessionCtx *common.Session) (res bool) {
+	if e.awsIAMAuthSupported != nil {
+		return *e.awsIAMAuthSupported
+	}
+	defer func() {
+		// cache result to avoid API calls on each new instance connection.
+		e.awsIAMAuthSupported = &res
+	}()
+	// check if the db supports IAM auth. If we get an error, assume the db does
+	// support IAM auth. False positives just incur an extra API call.
+	if ok, err := checkDBSupportsIAMAuth(sessionCtx.Database); err != nil {
+		e.Log.WithError(err).Debugf("Assuming database %s supports IAM auth.",
+			sessionCtx.Database.GetName())
+	} else if !ok {
+		return false
+	}
+	awsMeta := sessionCtx.Database.GetAWS()
+	dbUser := sessionCtx.DatabaseUser
+	ok, err := checkUserIAMAuthIsEnabled(ctx, e.CloudClients, awsMeta, dbUser)
+	if err != nil {
+		e.Log.WithError(err).Debugf("Assuming IAM auth is not enabled for user %s.",
+			dbUser)
+		return false
+	}
+	return ok
+}
+
+// checkDBSupportsIAMAuth returns whether the given database supports IAM auth.
+// AWS ElastiCache Redis supports IAM auth for redis version 7+.
+func checkDBSupportsIAMAuth(database types.Database) (bool, error) {
+	if !database.IsElastiCache() {
+		return false, nil
+	}
+	version, ok := database.GetLabel("engine-version")
+	if !ok {
+		return false, trace.NotFound("database missing engine-version label")
+	}
+	v, err := semver.NewVersion(strings.TrimPrefix(version, "v"))
+	if err != nil {
+		return false, trace.Wrap(err, "failed to parse engine-version")
+	}
+	return v.Major >= 7, nil
+}
+
+// checkUserIAMAuthIsEnabled returns whether a given ElastiCache user has IAM auth
+// enabled.
+func checkUserIAMAuthIsEnabled(ctx context.Context, clients cloud.Clients, awsMeta types.AWS, username string) (bool, error) {
+	client, err := clients.GetAWSElastiCacheClient(ctx, awsMeta.Region,
+		cloud.WithAssumeRoleFromAWSMeta(awsMeta))
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	// For IAM-enabled ElastiCache users, the username and user id properties
+	// must be identical.
+	// https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/auth-iam.html#auth-iam-limits
+	input := elasticache.DescribeUsersInput{UserId: aws.String(username)}
+	out, err := client.DescribeUsersWithContext(ctx, &input)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	if out == nil || len(out.Users) < 1 || out.Users[0].Authentication == nil {
```

Comment:
Does the out == nil && err != nil is a valid case ?  
```suggestion
	if len(out.Users) < 1 || out.Users[0].Authentication == nil {
```

Reply:
looking at AWS sdk code, it appears the output is never nil. I checked for nil here to be careful to avoid a panic if that output is ever nil with nil error, but I think it's safe to remove it.